### PR TITLE
ITP-JAN-LONDON |FOROGH AGHAEIYARIJANI |MODULE-DATA-GROUPS|SPRINT1|WEEK1|

### PR DESCRIPTION
**Pull request by Forogh-Aghaeiyarijani**
## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [x] My changes meet the [requirements](./README.md) of this task

## Changelist

Briefly explain your PR.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-1/fix/median.js [hunk 1]
```diff
@@ -3,9 +3,14 @@
 // If you're in the Sprint-1 directory, you can run `npm test -- fix` to run the tests in the fix directory
 
 function calculateMedian(list) {
-  const middleIndex = Math.floor(list.length / 2);
-  const median = list.splice(middleIndex, 1)[0];
-  return median;
+  const sortedList = [...list].sort((a, b) => a - b);
```
Comment (illicitonion) [inline Sprint-1/fix/median.js:6]: This looks like it works, but just to check I understand why you've done it this way - what's the benefit of using array destructuring here? What would happen if we didn't use it? Is there a test we could write to show it's needed?

File: Sprint-1/fix/median.js [hunk 2]
```diff
@@ -3,9 +3,14 @@
...
+  const middleIndex = Math.floor(sortedList.length / 2);
+
+  if (sortedList.length % 2 === 0) {
```
Comment (illicitonion) [inline Sprint-1/fix/median.js:9]: What would happen if we were passed an empty array? What answer would we return? What would we expect?

File: Sprint-1/implement/dedupe.test.js [hunk 3]
```diff
@@ -25,3 +25,21 @@ test.todo("given an empty array, it returns an empty array");
 // Given an array with strings or numbers
 // When passed to the dedupe function
 // Then it should remove the duplicate values, preserving the first occurence of each element
+const dedupe = require("./dedupe.js");
```
Comment (illicitonion) [inline Sprint-1/implement/dedupe.test.js:28]: It looks like you commented out an identical line on line 1 - why did you comment and re-add the same thing?

File: Sprint-1/implement/dedupe.test.js [hunk 4]
```diff
@@ -25,3 +25,21 @@ test.todo("given an empty array, it returns an empty array");
...
+
+describe("dedupe", () => {
+  test("given an empty array, it returns an empty array", () => {
+    expect(dedupe([])).toEqual([]);
+  });
+
+  test("given an array with no duplicates, it returns a copy of the original array", () => {
```
Comment (illicitonion) [inline Sprint-1/implement/dedupe.test.js:35]: You say in the test name "a copy", but I don't see the test testing that you actually copied the array, it could be from this test that the same array was returned as was passed in. How could you show this isn't happening in the test?

File: Sprint-1/implement/max.test.js [hunk 5]
```diff
@@ -41,3 +41,34 @@ test.todo("given an empty array, returns -Infinity");
 // Given an array with only non-number values
 // When passed to the max function
 // Then it should return the least surprising value given how it behaves for all other inputs
+const findMax = require("./max.js");
+
+describe("findMax", () => {
```
Comment (illicitonion) [inline Sprint-1/implement/max.test.js:46]: This is a great set of tests - it seem to cover lots of edge-cases, good job!

**Review by illicitonion — COMMENTED** [review]
This is generally looking really good - I've left a few questions about some generally quite small details :)

File: Sprint-1/fix/median.js (hunk 1 above)
Comment (Forogh-Aghaeiyarijani) [reply Sprint-1/fix/median.js:6]: This creates a shallow copy of the input array using array destructuring ([...]). The benefit of doing this is that it prevents the original array from being modified when you call sort().
The sort() method would modify the original list in place because sort() mutates the array it’s called on. This could lead to unexpected side effects if other parts of the code rely on the original order of the input array.
this test:
test("doesn't modify the input", () => {
  const list = [1, 2, 3];
  calculateMedian(list);

  expect(list).toEqual([1, 2, 3]);
});
If you remove the array destructuring and let sort() mutate the original array, this test will fail because the original list will no longer be [1, 2, 3] after calling calculateMedian().

File: Sprint-1/fix/median.js (hunk 2 above)
Comment (Forogh-Aghaeiyarijani) [reply Sprint-1/fix/median.js:9]: Return null or undefined to signal that the input is invalid.
Or throw an error (like new Error("Cannot calculate median of an empty array")).

File: Sprint-1/implement/dedupe.test.js (hunk 3 above)
Comment (Forogh-Aghaeiyarijani) [reply Sprint-1/implement/dedupe.test.js:28]: I temporarily commented it out while organizing the code to keep everything neat and aligned. Once I finished, I re-added it to make sure everything works correctly.

File: Sprint-1/implement/dedupe.test.js (hunk 4 above)
Comment (Forogh-Aghaeiyarijani) [reply Sprint-1/implement/dedupe.test.js:35]: To test that a new array (not the same reference) is returned, you can use toBe
I did it

Comment (illicitonion) [reply Sprint-1/implement/dedupe.test.js:35]: Nice, this is really good!

File: Sprint-1/fix/median.js (hunk 2 above)
Comment (illicitonion) [reply Sprint-1/fix/median.js:9]: Are you saying this is what you _expect_, or this is what _actually happens_?

Maybe you could write a test to make sure it's doing what you expect?

**Review by illicitonion — COMMENTED** [review]
Good answers, thanks!

File: Sprint-1/fix/median.js (hunk 2 above)
Comment (Forogh-Aghaeiyarijani) [reply Sprint-1/fix/median.js:9]: I did it.

**Review by illicitonion — COMMENTED** [review]
Perfect! Thanks so much, and congratulations!
